### PR TITLE
Add doc artifact

### PR DESCRIPTION
**Pull request by antisaling**
## Description
Adds a new artifact type called `DocArtifact` that should serve as a generic document artifact type.

## How Has This Been Tested?

Tested using `make test`.

## Merge criteria:

- [x] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: api/openapi/model-registry.yaml [hunk 1]
```diff
@@ -367,7 +367,7 @@ paths:
         "500":
           $ref: "#/components/responses/InternalServerError"
       operationId: getModelVersionArtifacts
-      summary: List All ModelVersion's artifacts
+      summary: List all artifacts
```
Comment (lampajr) [inline api/openapi/model-registry.yaml]: Why this change? this is actually the list of all `ModelVersion's artifacts` and not the list of all artifacts, isn't it?

**Review by lampajr — CHANGES_REQUESTED** [review]
Hey @isinyaaa ,

Thanks for this change!!!

I did not go through the whole change yet but I see there is some issue retrieving the Artifacts of a specific ModelVersion.

The issue I think is in the `GetModelVersionArtifacts` function which assumes all artifacts are `ModelArtifact` which is no more the case.

In order to reproduce the issue, I registered a model and a version and then I associated a DocArtifact to that version and issueing:
```
curl -X 'GET' \
  'http://localhost:8080/api/model_registry/v1alpha1/model_versions/2/artifacts?pageSize=100&orderBy=ID&sortOrder=DESC' \
  -H 'accept: application/json'
```

I get
```
{
  "message": "invalid entity: expected odh.ModelArtifact but received odh.DocArtifact, please check the provided id"
}
```

File: api/openapi/model-registry.yaml (hunk 1 above)
Comment (antisaling) [reply api/openapi/model-registry.yaml]: MB, I misread it as `ModelArtifact` 😅

**Review by lampajr — COMMENTED** [review]
Overall looks good to me, I tried to add some `DocArtifact` along with `ModelArtifact` and the requests worked fine:

Retrieve all artifacts associated to a model version:
```
curl --silent -X 'GET' 
 'http://localhost:8080/api/model_registry/v1alpha1/model_versions/2/artifacts?pageSize=100&orderBy=ID&sortOrder=DESC' \
  -H 'accept: application/json' | jq
{
  "items": [
    {
      "artifactType": "model-artifact",
      "createTimeSinceEpoch": "1707299253225",
      "customProperties": {},
      "description": "Version model",
      "id": "2",
      "lastUpdateTimeSinceEpoch": "1707299253225",
      "name": "mymodel",
      "state": "UNKNOWN",
      "uri": "http://path/to/model"
    },
    {
      "artifactType": "doc-artifact",
      "createTimeSinceEpoch": "1707299235321",
      "customProperties": {},
      "description": "Version README",
      "id": "1",
      "lastUpdateTimeSinceEpoch": "1707299235321",
      "name": "README",
      "state": "UNKNOWN",
      "uri": "http://path/to/README"
    }
  ],
  "nextPageToken": "",
  "pageSize": 100,
  "size": 2
}
```

Retrieve only model artifacts
```
curl --silent -X 'GET' 'http://localhost:8080/api/model_registry/v1alpha1/model_artifacts?pageSize=100&orderBy=ID&sortOrder=DESC' \ 
  -H 'accept: application/json' | jq
{
  "items": [
    {
      "artifactType": "model-artifact",
      "createTimeSinceEpoch": "1707299253225",
      "customProperties": {},
      "description": "Version model",
      "id": "2",
      "lastUpdateTimeSinceEpoch": "1707299253225",
      "name": "mymodel",
      "state": "UNKNOWN",
      "uri": "http://path/to/model"
    }
  ],
  "nextPageToken": "",
  "pageSize": 100,
  "size": 1
}
```

There are only a couple of points that in my opinion should be addressed, either in this PR or in a subsequent one:
* **Code coverage**: Improve a little bit the code coverage, based on the [report](https://github.com/opendatahub-io/model-registry/pull/289#issuecomment-1927257049) I think we could improve it a little bit --> I'd do this improvement in this PR directly if possible
* **Update DocArtifact**: I noticed that `DocArtifact` can be added but there is no way to update them once created, I think that it is worth to consider adding some endpoints for generic `get` and `patch` of a generic artifact, wdyt? - This can be done in a different PR if we agree

wdyt @isinyaaa @tarilabs

**Comment by antisaling** [comment]
> ## [Codecov](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io) Report
> Attention: `62 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`44ce399`)](https://app.codecov.io/gh/opendatahub-io/model-registry/commit/44ce3994fbd334bcd2d63c9ef432122b6005eff8?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io) 76.42% compared to head [(`e350114`)](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io) 75.60%.
> 
> [Files](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io)	Patch %	Lines
> [pkg/core/core.go](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io#diff-cGtnL2NvcmUvY29yZS5nbw==)	75.43%	[16 Missing and 12 partials ⚠️ ](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io)
> [internal/converter/openapi_mlmd_converter_util.go](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io#diff-aW50ZXJuYWwvY29udmVydGVyL29wZW5hcGlfbWxtZF9jb252ZXJ0ZXJfdXRpbC5nbw==)	19.23%	[20 Missing and 1 partial ⚠️ ](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io)
> [internal/mapper/mapper.go](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io#diff-aW50ZXJuYWwvbWFwcGVyL21hcHBlci5nbw==)	87.50%	[3 Missing and 2 partials ⚠️ ](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io)
> [internal/converter/mlmd_openapi_converter_util.go](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io#diff-aW50ZXJuYWwvY29udmVydGVyL21sbWRfb3BlbmFwaV9jb252ZXJ0ZXJfdXRpbC5nbw==)	55.55%	[3 Missing and 1 partial ⚠️ ](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io)
> [internal/converter/openapi_converter_util.go](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io#diff-aW50ZXJuYWwvY29udmVydGVyL29wZW5hcGlfY29udmVydGVyX3V0aWwuZ28=)	0.00%	[4 Missing ⚠️ ](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/opendatahub-io/model-registry/pull/289?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opendatahub-io).

When I test it locally I get some different results, wonder why that might be the case 🤔 

image

Looking at the report, I'd argue that we might want to merge a few tests in `core_test.go` as `ModelArtifact` and `DocArtifact` share much functionality.

**Review by lampajr — DISMISSED** [review]
>I'd argue that we might want to merge a few tests in core_test.go as ModelArtifact and DocArtifact share much functionality.

yeah I agree that we could think about refactoring the `core_test` a little bit as it is increasing more and more :)

Anyway the PR looks good to me, thx a lot @isinyaaa !

I created a follow-up [issue](https://github.com/opendatahub-io/model-registry/issues/297) with regard to the possibility to update generic `Artifact` records

**Review by tarilabs — APPROVED** [review]
Thank you so much @isinyaaa for this contribution and thank you @lampajr for taking a deep look into impl details!

Just seized the chance for a brief User Story coverage E2E and smoke testing the REST api layer respond, commit 6ed75c0077cdd16e1127a0c9d0216869abb76338

**Review by lampajr — APPROVED** [review]
Thanks @tarilabs for the addition, good to keep `robot` tests updated!

**Pull request merged**